File: android/src/main/java/com/aptoslabs/arculus/csdk/WalletVerifyPIN.kt
package com.aptoslabs.arculus.csdk

import com.sun.jna.Pointer

class WalletVerifyPIN(wallet: Pointer, private val pin: String) : CSDKAPICall<Unit>(wallet) {
  private class WalletVerifyPINException(tries: Int) :
    Exception("Wrong PIN. $tries tries remaining.")

  override suspend fun request(): Array<ByteArray> {
    val pinBytes = pin.encodeToByteArray()

    val len = SizeTByReference()

    val pointer = CSDK.WalletVerifyPINRequest(wallet, pinBytes, pinBytes.size, len)

    if (pointer == Pointer.NULL) throw CSDKAPICallRequestCreationException()

    return arrayOf(pointer.getByteArray(0, len.toIntChecked()))
  }

  override suspend fun response(bytes: ByteArray) {
    val nbrOfTries = SizeTByReference()

    val rc = CSDK.WalletVerifyPINResponse(wallet, bytes, bytes.size, nbrOfTries)

    val tries = nbrOfTries.toIntChecked()

    if (tries in 0..2) throw WalletVerifyPINException(tries)

    validateWalletResponse(rc)
  }
}

File: ios/Arculus/CSDK/WalletVerifyPIN.swift
import CSDK

final class WalletVerifyPIN: CSDKAPICall<Void> {
    enum VerifyPINError: Error, LocalizedError {
        case wrongPIN(Int)
        
        var errorDescription: String? {
            switch self {
            case let .wrongPIN(tries): return "Wrong PIN. \(tries) tries remaining."
            }
        }
    }
    
    private let pin: String
    
    init(wallet: OpaquePointer, pin: String) {
        self.pin = pin
        
        super.init(wallet: wallet)
    }
    
    override func request() async throws -> [Data] {
        var len: size_t = 0
        
        guard let bytes = WalletVerifyPINRequest(wallet, pin, pin.count, &len) else {
            throw CSDKAPICallError(.requestCreationFailed("verify PIN"))
        }
        
        return [Data(bytes: bytes, count: len)]
    }
    
    override func response(bytes: [UInt8]) async throws {
        var nbrOfTries: size_t = 0
        
        let code = WalletVerifyPINResponse(wallet, bytes, bytes.count, &nbrOfTries)
        
        if 0...2 ~= nbrOfTries {
            throw VerifyPINError.wrongPIN(nbrOfTries)
        }
        
        try CSDKAPICallError.validateWalletResponseCode(rc: code)
    }
}
